### PR TITLE
Add format validation for invoice attributes

### PR DESCRIPTION
**Pull request by mikaellisselgard**
PayEx have validations for the invoice attributes, we reflect those
validations here to avoid sending invalidig attributes to PayEx. This
patch also introduces a resolve option, where you can pass a Proc to try
to format the attribute correctly and retry the validation. For example
shorten the customer name if it's too long. See below for validations:

http://www.payexpim.com/technical-reference/pxorder/purchaseinvoiceprivate/

### CODE REVIEW

File: lib/payex/api/pxorder.rb [hunk 1]
```diff
@@ -103,65 +103,83 @@ def PurchaseInvoicePrivate(params)
...
       },
       'country' => {
-        signed: true
+        signed: true,
+        format: /[A-Z]{,2}$/
       },
       'socialSecurityNumber' => {
-        signed: true
+        signed: true,
+        format: /^[a-zA-Z0-9_:!;\\”#<>=?\\[\\]@{}´\n\r %-\/À-ÖØ-öø-úü]{,15}$?/
```
Comment (hovsater) [inline lib/payex/api/pxorder.rb]: I'm not entirely sure this regex is correct. Try this instead `\A[a-zA-Z0-9_:!;\"#<>=?\\[\\]@{}´\n\r %-\/À-ÖØ-öø-úü]*\z`.

File: lib/payex/api.rb [hunk 2]
```diff
@@ -110,4 +112,16 @@ def parse_transaction_status(status)
     else status.to_s
     end
   end
+
+  def valid_param_format?(param, options)
+    !options.include?(:format) || options[:format] === param
+  end
+
+  def parse_with_resolve!(param, options)
+    resolve = options[:resolve]
```
Comment (hovsater) [inline lib/payex/api.rb]: Could also be written as:

```ruby
resolve = options.fetch(:resolve) { param_error! 'resolve parameter required' }
```

File: lib/payex/api.rb [hunk 3]
```diff
@@ -110,4 +112,16 @@ def parse_transaction_status(status)
...
+    param_error! %{resolve parameter required} if resolve.nil?
+    result = resolve.call(param) if resolve.is_a? Proc
```
Comment (hovsater) [inline lib/payex/api.rb]: I think asking if it responds to `#call` is better. What do you think?

**Comment by hovsater** [comment]
Could also be written as:

```ruby
def parse_param(param, options)
  value = param || options.fetch(:default) { param_error! 'parameter required' }
  value = value.call if value.respond_to?(:call)
  
  return value if options[:format] == value
  
  if is_resolver?(options[:resolve])
    parse_param(options[:resolver].call(value), options.except(:default, :resolve))
  else
    param_error! %{must match #{options[:format].inspect}}
  end
end


def is_resolver?(resolver)
  resolver.respond_to?(:call)
end
```

File: lib/payex/api/pxorder.rb [hunk 4]
```diff
@@ -103,65 +103,83 @@ def PurchaseInvoicePrivate(params)
...
       'customerRef' => {
-        signed: true
+        signed: true,
+        format: /\A[a-zA-Z0-9]{,15}\z/
       },
       'customerName' => {
-        signed: true
+        signed: true,
+        format: /\A[a-zA-Z0-9_:!;\"#<>=?\\[\\]@{}´\n\r %-\/À-ÖØ-öø-úü]{,35}\z/,
+        resolve: proc { |v| v[0..34].strip }
```
Comment (hovsater) [inline lib/payex/api/pxorder.rb]: could also be written as `-> (v) { }`

File: spec/api_spec.rb [hunk 5]
```diff
@@ -57,4 +57,115 @@
...
       should == { 'a' => 1, 'b' => 2 }
   end
+
+  describe '#parse_param' do
+    context 'when format validation fails' do
+      context 'when resolve proc is present' do
+        it 'calls the proc' do
+          options = {
+            format: /^[a-z]{,3}$/,
+            resolve: proc { |v| v[0..2].strip }
```
Comment (hovsater) [inline spec/api_spec.rb]: Extract the proc to a variable, and the use it in the options. It reads better.

```ruby
resolver = -> (v) { v[0..2].strip }

options = { format: /^[a-z]{,3}$/, resolve: resolver }

resolver.should_receive(:call).with('too long param').and_return('too')
```

**Review by hovsater — CHANGES_REQUESTED** [review]

**Comment by mikaellisselgard** [comment]
@KevinSjoberg Vettiga förslag, fixat här: https://github.com/standout/payex-ruby/pull/4/commits/88c99e5fc0cc7b9d57d8279781872097d0e46ac2

**Review by hovsater — APPROVED** [review]

File: spec/api_spec.rb [hunk 6]
```diff
@@ -57,4 +57,85 @@
     PayEx::API.parse_params({ a: 1 }, { 'b' => { default: 2 } }).
       should == { 'a' => 1, 'b' => 2 }
   end
+
+  describe '#parse_param' do
+    context 'when format validation fails' do
+      context 'when resolve proc is present' do
+        it 'calls the proc' do
```
Comment (Istanful) [inline spec/api_spec.rb:64]: Why shoulda matchers?

Provided you can I would probably write something like:
```ruby
resolver = -> (v) { v[0..2].strip }
options = { format: /^[a-z]{,3}$/, resolve: resolver }
allow(resolver).to receive(:call).and_return('too')

PayEx::API.parse_param('too long param', options)

expect(resolver).to have_received(:call).with('too long param')
```

**Review by Istanful — APPROVED** [review]
Fix if possible and you want. Good otherwise 👍

File: spec/api_spec.rb (hunk 6 above)
Comment (mikaellisselgard) [reply spec/api_spec.rb:64]: It's because we're running rspec 2.11.

Comment (Istanful) [reply spec/api_spec.rb:64]: Gotcha. I suspected something like that. Feel free to merge :)

**Pull request merged**